collector/supervisord.go

```diff
@@ -0,0 +1,142 @@
+)
+
+var (
+	supervisordURL = flag.String("collector.supervisord.url", "http://localhost:9001/RPC2", "XML RPC endpoint")
+)
+
+type supervisordCollector struct {
+	client     *xmlrpc.Client
+	state      *prometheus.GaugeVec
+	exitStatus *prometheus.GaugeVec
+	uptime     *prometheus.CounterVec
+}
+
+func init() {
+	Factories["supervisord"] = NewSupervisordCollector
+}
+
+func NewSupervisordCollector() (Collector, error) {
+
+	client, err := xmlrpc.NewClient(*supervisordURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		namespace  = "supervisor"
+		subsystem  = "service"
+		labelNames = []string{"name", "group"}
+	)
+	return &supervisordCollector{
+		client: client,
+		state: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "state",
+				Help:      "Process State",
+			},
+			labelNames,
+		),
+		exitStatus: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "exit_status",
+				Help:      "Process Exit Statue",
+			},
+			labelNames,
+		),
+		uptime: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "uptime",
+				Help:      "Process Uptime",
+			},
+			labelNames,
+		),
+	}, nil
+}
+
+func (c *supervisordCollector) isRunning(state int) bool {
+	// http://supervisord.org/subprocess.html#process-states
+	const (
+		STOPPED  = 0
+		STARTING = 10
+		RUNNING  = 20
+		BACKOFF  = 30
+		STOPPING = 40
+		EXITED   = 100
+		FATAL    = 200
+		UNKNOWN  = 1000
+	)
+	switch state {
+	case STARTING:
+	case RUNNING:
+	case STOPPING:
+	case BACKOFF:
+		return true
+	}
+	return false
+}
+
+func (c *supervisordCollector) Update(ch chan<- prometheus.Metric) error {
+
+	var infos []struct {
+		Name          string `xmlrpc:"name"`
+		Group         string `xmlrpc:"group"`
+		Start         int    `xmlrpc:"start"`
+		Stop          int    `xmlrpc:"stop"`
+		Now           int    `xmlrpc:"now"`
+		State         int    `xmlrpc:"state"`
+		StateName     string `xmlrpc:"statename"`
+		SpawnErr      string `xmlrpc:"spanerr"`
+		ExitStatus    int    `xmlrpc:"exitstatus"`
+		StdoutLogfile string `xmlrcp:"stdout_logfile"`
+		StderrLogfile string `xmlrcp:"stderr_logfile"`
+		PID           int    `xmlrpc:"pid"`
+	}
+	if err := c.client.Call("supervisor.getAllProcessInfo", nil, &infos); err != nil {
+		return err
+	}
+	for _, info := range infos {
+		lables := []string{info.Name, info.Group}
+		c.state.WithLabelValues(lables...).Set(float64(info.State))
+		c.exitStatus.WithLabelValues(lables...).Set(float64(info.ExitStatus))
+		if c.isRunning(info.State) {
+			c.uptime.WithLabelValues(lables...).Set(float64(info.Now - info.Start))
+		} else {
+			c.uptime.WithLabelValues(lables...).Set(0)
+		}
+		log.Debugf("%s:%s is %s on pid %d", info.Group, info.Name, info.StateName, info.PID)
+	}
+
+	c.state.Collect(ch)
+	c.exitStatus.Collect(ch)
+	c.uptime.Collect(ch)
```

Comment:
Hm, wouldn't it necessary to reset the gauges at some point, in order to not export stale information? Or am I missing that somewhere?

Alternatively, you could take look at the low level form and update the channel directly with a metric value without storing it in a gauge collector first. Here is an example: https://github.com/prometheus/node_exporter/blob/9810c57725f72ea32e73b2e4c74800f50a6328ef/collector/uname_linux.go#L55